pkg/registry/core/service/storage/storage.go

```diff
@@ -192,7 +193,8 @@ func (r *GenericREST) defaultSvcOnRead(service *api.Service) error {
 			}
 		} else {
 			// headless w/ selector
-			// follows cluster configuration
+			// follows cluster configuration. this service (selector based) uses a selector and will have to follow
+			// how the cluster is configured, if it is configured to dual stack then the service must be dualstacked.
 			if r.secondaryFamily != nil { // dualstack cluster
 				service.Spec.IPFamilies = append(service.Spec.IPFamilies, *r.secondaryFamily)
 				service.Spec.IPFamilyPolicy = &requireDualStack
```

Comment:
I am still unclear why REQUIRED vs PREFER here?  In fact, why not use prefer for any headless service which does not specify.  That way, if a cluster is upgraded to dual, then downgraded this field will stay constant (though ipFamilies can change).

Otherwis this value will change from single->dual->single

Reply:
I had suggested in another comment that we should revisit the naming of the `ipFamilyPolicy` values (after this PR has merged) because the "Require" vs "Prefer" distinction doesn't really make sense for headless services the way we have defined things: a headless service is allowed to force itself to be dual-stack regardless of the cluster configuration, so it has no need to express a "preference" for being dual stack, and it has no need to "require" that the rest of the cluster be dual stack. It's not "preferring" or "requiring" dual stack, it just _is_ dual stack.

So I think we should rename `RequireDualStack` to just `DualStack` and then say:

1. Headful services would still have the same semantics they have now:
   - `ipFamilyPolicy: SingleStack` means it is single stack regardless of configuration, just like now
   - `ipFamilyPolicy: PreferDualStack` means it can be single stack or dual stack depending on configuration, just like now
   - `ipFamilyPolicy: DualStack` means it can only be dual stack, just like `ipFamilyPolicy: RequireDualStack` means now
2. Headless services would have slightly clearer semantics:
   - `ipFamilyPolicy: SingleStack` means it only wants endpoints of `ipFamilies[0]`, and CoreDNS, etc, should ignore any endpoints of the other family. This is the default for headless services with selectors in single-stack clusters.
   - `ipFamilyPolicy: DualStack` means it is _allowed_ to have endpoints of both families. It doesn't necessarily mean it _will_ have endpoints of both families, but it's allowed to. This is the default for headless services with selectors in dual-stack clusters, and for headless selectorless services in all clusters.
   - `ipFamilyPolicy: PreferDualStack` would not be allowed for headless services, because it has no meaning